cs/src/core/Serializer.cs

```diff
@@ -76,10 +77,23 @@ public class Serializer<W>
         public Serializer(Type type, bool inlineNested)
         {
             var parser = new ObjectParser(type);
-            serialize = SerializerGeneratorFactory<object, W>.Create(
-                    (o, w, i) => serialize[i](o, w), type, inlineNested)
-                .Generate(parser)
-                .Select(lambda => lambda.Compile()).ToArray();
+
+            var firstPassAttribute = typeof(W).GetAttribute<FirstPassWriterAttribute>();
+            if (firstPassAttribute != null)
+            {
+                if (!typeof(ITwoPassProtocolWriter).IsAssignableFrom(typeof(W)))
+                {
+                    throw new ArgumentException("Writers with FirstPassWriterAttribute must implement ITwoPassProtocolWriter");
+                }
+
+                Type firstPassHelperType = typeof(TwoPassSerializerHelper<>).MakeGenericType(typeof(W), firstPassAttribute.Type);
+
+                helper = (SerializerHelper)Activator.CreateInstance(firstPassHelperType, parser, type, inlineNested);
+            }
+            else
+            {
+                helper = new SerializerHelper(parser, type, inlineNested);
+            }
```

Comment:
1.  Not sure how to cleanly merge Serializer&Transcoder, 'cause they use different helper types, have different # of generic parameters, and their constructors take different args.
2.  Done.